### PR TITLE
Add Mongo code lens provider, Run, Run All, show connected DB and show commands in output

### PR DESCRIPTION
**Pull request by StephenWeatherford**
1) Show executing command in output
2) Run All
3) Code Lens for "Connected to <db>", "Run All" and "Run"
4) Execute Command -> Execute Mongo Command (editor context menu)
5) Commands.ts -> MongoCommands.ts

image

image

image

### CODE REVIEW

**Comment by StephenWeatherford** [comment]
@fiveisprime 
Question: Should it be "Execute" and "Execute All" to match the current command names?

File: src/mongo/MongoCommands.ts [hunk 1]
```diff
@@ -18,16 +18,61 @@ import { MongoFindOneResultEditor } from './editors/MongoFindOneResultEditor';
...
+		const commands = MongoCommands.getAllCommandsFromActiveEditor();
 		const activeEditor = vscode.window.activeTextEditor;
-		if (activeEditor.document.languageId !== 'mongo') {
-			vscode.window.showInformationMessage("This command can only be run inside of a MongoDB scrapbook (*.mongo)");
-			return;
-		}
 		const selection = activeEditor.selection;
-		const command = MongoCommands.getCommand(activeEditor.document.getText(), selection.start);
+		const command = MongoCommands.findCommandAtPosition(commands, selection.start);
+		return this.executeCommand(activeEditor, database, extensionPath, editorManager, tree, context, command);
```
Comment (PrashanthCorp) [inline src/mongo/MongoCommands.ts]: 1. The order of these arguments is slightly different from the executeCommands function. Would it make sense to make these consistent? 
2. Why is this returned when the other use case (inside executeCommands) awaited? 

As an aside, would it make sense to just have one function, `executeCommands`, with this specific case having a function call - `await MongoCommands.executeCommands([command], vscode.window.activeTextEditor, database, extensionPath, editorManager, tree, context)` where we pass the command in an array? I think there might be lesser code, but it could seem less readable.

**Comment by StephenWeatherford** [comment]
The run button always runs the line of code beneath it.  That's how all similar features I've been work, I don't think that will be confusing.

File: src/extension.ts [hunk 2]
```diff
@@ -37,6 +38,9 @@ export function activate(context: vscode.ExtensionContext) {
 	context.subscriptions.push(tree);
 	context.subscriptions.push(vscode.window.registerTreeDataProvider('cosmosDBExplorer', tree));
 
+	const codeLensProvider = new MongoCodeLensProvider();
+	context.subscriptions.push(vscode.languages.registerCodeLensProvider('mongo', codeLensProvider));
+
```
[the remark this answers is not in the document]
Comment (StephenWeatherford) [reply src/extension.ts]: done

File: src/mongo/MongoCommands.ts (hunk 1 above)
Comment (StephenWeatherford) [reply src/mongo/MongoCommands.ts]: I prefer the readability

File: src/mongo/MongoCommands.ts [hunk 3]
```diff
@@ -18,16 +18,61 @@ import { MongoFindOneResultEditor } from './editors/MongoFindOneResultEditor';
...
+
+	public static async executeCommands(activeEditor: vscode.TextEditor, database: IAzureParentNode<MongoDatabaseTreeItem>, extensionPath, editorManager: CosmosEditorManager, tree: AzureTreeDataProvider, context: IActionContext, commands: MongoCommand[]): Promise<void> {
+		for (let command of commands) {
+			await this.executeCommand(activeEditor, database, extensionPath, editorManager, tree, context, command);
+		}
+	}
+
+	public static async executeCommand(activeEditor: vscode.TextEditor, database: IAzureParentNode<MongoDatabaseTreeItem>, extensionPath, editorManager: CosmosEditorManager, tree: AzureTreeDataProvider, context: IActionContext, command: MongoCommand): Promise<void> {
 		if (command) {
+			output.show();
```
Comment (EricJizbaMSFT) [inline src/mongo/MongoCommands.ts]: Not a fan of this output - did you add it for debugging purposes only? A few reasons:
1. The user should know what command we're running and it feels like a bug if we're not running the command they expect
1. There's no way to link this output to its corresponding result.json file
1. We're already popping up the result.json file and I think popping up the output channel as well is too much

I guess I'm okay with this as long as you remove `output.show`.

Another idea is to insert the command to the top of the result.json file as a comment (although comments aren't strictly supported in json files so idk)

File: src/mongo/registerMongoCommands.ts [hunk 4]
```diff
@@ -94,29 +98,49 @@ export function registerMongoCommands(context: vscode.ExtensionContext, actionHa
...
-            const persistedNode: IAzureNode | undefined = await tree.findNode(persistedNodeId);
-            if (persistedNode) {
-                await languageClient.client.onReady();
-                await vscode.commands.executeCommand('cosmosDB.connectMongoDB', persistedNode);
-            }
+            persistedNode = await tree.findNode(persistedNodeId);
         }
     });
+
+    if (persistedNode) {
```
Comment (EricJizbaMSFT) [inline src/mongo/registerMongoCommands.ts]: Why did you move this code outside of `callWithTelemetryAndErrorHandling`? It was put in there since we await 'loadPersistedMongoDBTask' multiple times and we don't want the error being displayed multiple times

File: src/mongo/services/MongoCodeLensProvider.ts [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+				command: isInitialized && 'cosmosDB.connectMongoDB'
+			},
+			range: new vscode.Range(new vscode.Position(0, 0), new vscode.Position(0, 0))
+		});
+
+		if (isConnected) {
+			// Run all
+			lenses.push(<vscode.CodeLens>{
+				command: {
+					title: "Run All",
```
Comment (EricJizbaMSFT) [inline src/mongo/services/MongoCodeLensProvider.ts]: Can we pick either 'Execute' or 'Run'? I don't have a preference, but it would be nice to be consistent

File: package.json [hunk 6]
```diff
@@ -113,7 +115,12 @@
 			{
 				"category": "MongoDB",
 				"command": "cosmosDB.executeMongoCommand",
-				"title": "Execute Command"
+				"title": "Execute Mongo Command"
```
[the remark this answers is not in the document]
Comment (StephenWeatherford) [reply package.json]: The Mongo/MongoDB is inconsistent.  For instance, it's "New Mongo Scrapbook", and the name of the "language" is currently "mongo", and the extension is "*.mongo".  I would have chosen MongoDB, but not sure what it should be now.

Added Mongo to the command because it's not very clear that the context menu in the scrapbook refers to this.  There can be lots of other commands in there, too, and simply "Execute Command" is rather vague (I passed it up the first time I saw it).

image

File: src/mongo/MongoCommands.ts [hunk 7]
```diff
@@ -18,16 +18,61 @@ import { MongoFindOneResultEditor } from './editors/MongoFindOneResultEditor';
...
+		return this.executeCommand(activeEditor, database, extensionPath, editorManager, tree, context, command);
+	}
+
+	public static getAllCommandsFromActiveEditor(): MongoCommand[] {
+		const activeEditor = vscode.window.activeTextEditor;
+		if (activeEditor) {
+			const commands = MongoCommands.getAllCommandsFromTextDocument(activeEditor.document);
+			return commands;
+		} else {
+			throw new Error(notInScrapbookMessage);
```
[the remark this answers is not in the document]
Comment (StephenWeatherford) [reply src/mongo/MongoCommands.ts]: This is currently there (not my code) because the commands are exposed in the command palette but are always enabled. Excluding them sounds good, except that would make it inaccessible unless I add a keyboard shortcut.

File: src/mongo/MongoCommands.ts (hunk 3 above)
Comment (StephenWeatherford) [reply src/mongo/MongoCommands.ts]: Personally I much prefer to see what it's doing, especially for Execute All, and if it hits an error.  Fine with not showing output, though.  We might need to refine this behavior in some way. We can revisit after people have played with it more.

File: src/mongo/registerMongoCommands.ts (hunk 4 above)
Comment (StephenWeatherford) [reply src/mongo/registerMongoCommands.ts]: Doesn't matter how many times it's awaited, it will only run once.  

CodeLens will show "Initializing" until told setConnectedDatabase called first time, so have to make sure that happens. However, current version doesn't handle errors connecting to DB, I think this next one will be better...

File: src/mongo/MongoCommands.ts [hunk 8]
```diff
@@ -18,16 +18,61 @@ import { MongoFindOneResultEditor } from './editors/MongoFindOneResultEditor';
 import { MongoCommand } from './MongoCommand';
 import { MongoDatabaseTreeItem } from './tree/MongoDatabaseTreeItem';
 
+const output = vscodeUtil.getOutputChannel();
+const notInScrapbookMessage = "This command can only be run inside of a MongoDB scrapbook (*.mongo)";
+
 export class MongoCommands {
+
+	public static async executeAllCommandsFromActiveEditor(database: IAzureParentNode<MongoDatabaseTreeItem>, extensionPath, editorManager: CosmosEditorManager, tree: AzureTreeDataProvider, context: IActionContext): Promise<void> {
```
[the remark this answers is not in the document]
Comment (EricJizbaMSFT) [reply src/mongo/MongoCommands.ts]: FWIW I agree it's generally better to keep functional changes separate from refactoring, but I make an exception when people clean up old code that they're actively changing anyways. It feels like a sustainable way to clean up code as we go rather than filing a bunch of engineering issues. No strong opinion, though - separate PR is fine.

File: src/mongo/MongoCommands.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (EricJizbaMSFT) [reply src/mongo/MongoCommands.ts]: I thought it was accessible through code lens?

File: src/mongo/MongoCommands.ts (hunk 3 above)
Comment (EricJizbaMSFT) [reply src/mongo/MongoCommands.ts]: Good point - but it feels like we should include the command in the error message anyways

File: src/mongo/MongoCommands.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (StephenWeatherford) [reply src/mongo/MongoCommands.ts]: "Accessible" accessible - ie through keyboard. Maybe code lens does have this, will have to look it up.  If there is, I have no strong opinion.

**Comment by fiveisprime** [comment]
The code lens text in a new file ends up out of view for me and I have to scroll down to see it. After I scroll it into view, typing shifts everything back up and out of view.

**Comment by StephenWeatherford** [comment]
@fiveisprime I was just noticing that, are you using insiders?

File: src/mongo/MongoCommands.ts (hunk 3 above)
Comment (StephenWeatherford) [reply src/mongo/MongoCommands.ts]: I still think we need to play with it and make holistic changes in the flow rather than dealing with it in this PR.  The error handling is not new behavior, although execute all is new.

File: src/mongo/MongoCommands.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (StephenWeatherford) [reply src/mongo/MongoCommands.ts]: Can't seem to find a shortcut for this in vscode.

**Comment by fiveisprime** [comment]
Yeah, on Insiders.

**Comment by StephenWeatherford** [comment]
Don't think I saw it at work with current, will try when I get in office.

File: src/mongo/MongoCommands.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (EricJizbaMSFT) [reply src/mongo/MongoCommands.ts]: Oh that "Accessible". Well I definitely think you should exclude `executeMongoCommand` from the command palette. You lose context of the cursor when you open the command palette so it always runs the first command. Plus it already has a keyboard shortcut.

I guess it might make sense to have `executeAllMongoCommands` in the command palette, but a keyboard shortcut feels more natural to me. No strong opinion on this one. I will point out that whether or not you add a default keyboard shortcut, the user can add a custom keyboard shortcut for literally any command we define.

File: src/mongo/MongoCommands.ts (hunk 3 above)
Comment (EricJizbaMSFT) [reply src/mongo/MongoCommands.ts]: That's fine - you already removed `output.show()` which was the "new" behavior that I cared about the most

File: src/mongo/MongoCommands.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (StephenWeatherford) [reply src/mongo/MongoCommands.ts]: @fiveisprime any suggestions on a keyboard shortcut for execute all?  
Currently CTRL+SHIFT+” is execute.  I would choose CTRL+SHIFT+ENTER but that’s Insert Line Above”.

File: src/mongo/registerMongoCommands.ts (hunk 4 above)
Comment (EricJizbaMSFT) [reply src/mongo/registerMongoCommands.ts]: If you await a promise multiple times, you're correct that it only runs once. However, if the result of awaiting the promise is an error, you will get that error thrown every single time you await it. For example you can test this out by putting `throw new Error('Hi Stephen!');` right before `codeLensProvider.setConnectedDatabase(undefined);`. The error will pop up every time you execute a mongo command (not just the first time).

Instead, I suggest moving everything inside `await callWithTelemetryAndErrorHandling(` and using a 'try/finally' to make sure CodeLens always gets initialized

Comment (StephenWeatherford) [reply src/mongo/registerMongoCommands.ts]: setConnectedDatabase can't fail, but yeah I agree it's better inside the telemetry.

**Comment by StephenWeatherford** [comment]
@fiveisprime https://github.com/Microsoft/vscode/issues/47703

**Comment by fiveisprime** [comment]
@StephenWeatherford 

1. I like `Execute` and `Execute All`
1. How about moving to the left instead of right for Execute All shortcut: `CTRL+SHIFT+:`
1. Thanks for creating that issue

File: src/mongo/registerMongoCommands.ts (hunk 4 above)
Comment (EricJizbaMSFT) [reply src/mongo/registerMongoCommands.ts]: "can't fail" <- famous last words 😂

Comment (StephenWeatherford) [reply src/mongo/registerMongoCommands.ts]: "As currently implemented" (tm)

**Comment by EricJizbaMSFT** [comment]
@StephenWeatherford if you're trying to fix the build, this is the error:
```
src/mongo/services/MongoCodeLensProvider.ts(22,58): error TS6133: 'token' is declared but its value is never read.
```

We have this known issue: https://github.com/Microsoft/vscode-cosmosdb/issues/408 I guess one solution is to add `npm run build` before `gulp package` in the travis config. Technically it's redundant since the package runs the build, but that way we actually get to see the error messages.
